### ISSUE TITLE
Question: base64 vs blob for small/medium size images

### ISSUE THREAD

**Issue by DZakh** [dexie/Dexie.js#1402]
In our project we use base64 for avatars and thumbnails to render them as fast as possible, but the original image comes as blob and it doesn't really matter when to convert it to base64. So we came across a question which format is better for storing in the database. It looks like base64 takes more disk space, but what about the speed? I've seen somewhere in the source code that blob gets reconstructed before writing it to the database (I'm not really sure), so it looks like base64 should be faster, but I'd like to know if you have better knowledge in this area. Thanks.

**Comment by tiptronic** [comment]
@DZakh Did you finally come to a conclusion what's better to use, in terms of speed (base64 vs. blob vs. arraybuffer)? 
In my case I'm also just storing lots of thumbnails (<200x200px)

**Comment by dfahlander** [comment]
> @DZakh Did you finally come to a conclusion what's better to use, in terms of speed (base64 vs. blob vs. arraybuffer)? In my case I'm also just storing lots of thumbnails (<200x200px)

I would use a Blob but as I said, I haven't done any performance measuring on this.

**Comment by dfahlander** [comment]
> @DZakh Did you finally come to a conclusion what's better to use, in terms of speed (base64 vs. blob vs. arraybuffer)? In my case I'm also just storing lots of thumbnails (<200x200px)

I would use a Blob but as I said, I haven't done any performance measuring on this.

**Comment by dfahlander** [comment]
> @DZakh Did you finally come to a conclusion what's better to use, in terms of speed (base64 vs. blob vs. arraybuffer)? In my case I'm also just storing lots of thumbnails (<200x200px)

I would use a Blob but as I said, I haven't done any performance measuring on this.

**Comment by tiptronic** [comment]
@dfahlander Thanks - and that's what I'm doing atm - I just didn't do some performance comparison, e.g. when trying to load/show a couple of hundereds of these blobs quickly (e.g. thumbnails)... 

Before that I used Arraybuffers (which also worked quite good) and switched to Blobs as they are a bit easier to handle (and I think it's a little bit better in terms of memory usage). But Dexie did handle all types fine so far.

**Comment by DZakh** [comment]
> @DZakh Did you finally come to a conclusion what's better to use, in terms of speed (base64 vs. blob vs. arraybuffer)? In my case I'm also just storing lots of thumbnails (<200x200px)

Yes, we've decided to store small pictures such as thumbnails in base64. The only advantage of base64 over blob is that it displayed synchronously. For example it's very important with infinite lists.  

Talking about performance there's almost no difference for images lighter than 1mb, but after this the processing time increases exponentially. I didn't make measurements with dexie, but I did with sending images to worker. For small files it was around 2ms both base64 and blob, but when I did measurements with bigger files around 50mb it became ~1500ms for base64, but stayed 2ms for blob.

Talking about size, base64 is around 1.5 times heavier than blob, but as I said before it's not a big difference for small images.

If you need synchronously display thumbnails or to store small images in some weird plaices (eg local storage), then stick with base64, in all other cases I would consider blob.  
It's not a question of performance but a problem you want to solve.

**Comment by tiptronic** [comment]
@DZakh Wow - thanks for this very valuable information 👍 
In my case I am loading a couple of thousand SVGs from the file-system (in the browser - using `fetch`-requests (or `xhr`) and show them in a grid (<li> + using some intersectionObserver to only show the visibles).

The process is pretty slow if I just set the` image.src` to the svg (because the svgs must be read, and rendered). So the idea is to save a small thumbnail (as `png`) of a rendered svg on first load and from there on show the png when rendering the list. This should be much quicker than showing the svg, especially if I can get the thumbnails out of the db quickly :) 

So for me both options will work - Blob or base64 - but if I read your comment correctly base64 could have some benefits for my scenario...

**Comment by DZakh** [comment]
If users scroll fast, they will see a blank space instead of a thumbnail, because it takes some time to load an url from blob (between 5ms-25ms in my case, you can check it in the network tab). It's only for a moment, but noticeble. You can only solve the problem with base64. Moreover even with base64 there's another problem - how to load the resource itself before it's shown to user :)

**Comment by tiptronic** [comment]
Haha - thanks - that's okay... I am showing a 1-px placeholder in the grid (scaled to target-size with css) while the image loads, so that's fine for me. I'll check that out asap, but your comments are really great - so I'll give base64 a whirl first...

**Comment by tiptronic** [comment]
@DZakh After some experiments I am facing another issue - so I hope you have some insights here as well:
- I am reading SVGs and converting them to base64 (image/png)
- I tried doing that as fast as possible, but it seems indexedDB doesn't like that (I tried `Dexie.put` and `Dexie.bulkPut`) but it seems indexedDB is not getting all images (records missing,etc...).

- @dfahlander Did you (or someone else) try to create 2000+ thumbnails fast to a Dexie database? Could it be that I need to throttle (or queue) lots of images? There are really very, very little places on the web where people store images (or even larger amount of images) in Dexie (most what I found is strings/json). So any comment welcome here.

- After creating a base64 image, would it make sense to create that image to a blob (with even more overhead), to make use of async loading and optimized memory?
